File: app/src/main/java/com/task/paginationlist/presentation/piclist/views/adapters/PicAdapter.java
package com.task.paginationlist.presentation.piclist.views.adapters;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.task.paginationlist.PaginationListApplication;
import com.task.paginationlist.R;
import com.task.paginationlist.data.db.models.WallpaperDb;
import com.task.paginationlist.data.utils.ILoader;
import com.task.paginationlist.presentation.piclist.interfaces.OnPickClickListener;
import com.task.paginationlist.presentation.piclist.views.activities.interfaces.IPaginatorAdapter;
import com.task.paginationlist.presentation.piclist.views.adapters.holders.LoadingHolder;
import com.task.paginationlist.presentation.piclist.views.adapters.holders.PicViewHolder;
import com.task.paginationlist.presentation.piclist.views.paginator.Config;

import java.text.ParseException;
import java.util.List;

public class PicAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements IPaginatorAdapter {
    private static final String TAG = "PicAdapter";
    public static final int ITEM_TYPE = 1;
    public static final int FOOTER_TYPE = 0;

    private boolean isSuccessful = true;

    private View.OnClickListener reload;
    private OnPickClickListener onPicClick;


    ILoader loader;

    private List<WallpaperDb> wallpapers;

    public PicAdapter() {
        loader = PaginationListApplication.getComponent().getLoader();
    }

    public void addData(List<WallpaperDb> models, boolean isRefreshing) {
        if (isRefreshing) {
            this.wallpapers = null;
        }
        if (this.wallpapers == null) {
            this.wallpapers = models;
        } else {
            this.wallpapers.addAll(models);
        }
        isSuccessful = true;
        notifyDataSetChanged();
    }

    public void setReload(View.OnClickListener listener) {
        this.reload = listener;
    }

    public void setOnPicClick(OnPickClickListener onPicClick) {
        this.onPicClick = onPicClick;
    }

    public void setSuccessful(boolean successful) {
        isSuccessful = successful;
        notifyItemChanged(getItemCount() - 1);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (isFooter(position)) {
            LoadingHolder loadingHolder = ((LoadingHolder) holder);
            loadingHolder.loaderView.setReloadClick(reload);
            if (!isSuccessful) {
                loadingHolder.loaderView.showError();
            } else {
                ((LoadingHolder) holder).loaderView.showBar();
            }
        } else {
            PicViewHolder picHolder = (PicViewHolder) holder;
            loader.loadImage(wallpapers.get(position).getThumbUrl(), picHolder.image);
            picHolder.image.setOnClickListener(v -> {
                if (onPicClick != null) {
                    onPicClick.onPickClick(v, position, wallpapers.get(position).getId(), wallpapers.size());
                }
            });
            String text = "";
            try {
                text = Config.formatDate(wallpapers.get(position).getCreationDate());
            } catch (ParseException e) {
                Log.e(TAG, e.getMessage());
            }
            picHolder.date.setText(text);
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return viewType == FOOTER_TYPE ? new LoadingHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.footer_item, parent, false))
                : new PicViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.pic_list_item, parent, false));
    }

    @Override
    public int getItemViewType(int position) {
        return isFooter(position) ? FOOTER_TYPE : ITEM_TYPE;
    }

    private boolean isFooter(int position) {
        return position == getItemCount() - 1;
    }

    public int getRealItemCount() {
        return getItemCount() - 1;
    }

    @Override
    public int getItemCount() {
        return wallpapers != null ? wallpapers.size() + 1 : 1;
    }
}

File: app/src/main/java/com/task/paginationlist/presentation/piclist/views/adapters/PicAdapter.kt
package com.task.paginationlist.presentation.piclist.views.adapters

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.task.paginationlist.PaginationListApplication
import com.task.paginationlist.R
import com.task.paginationlist.data.db.models.WallpaperDb
import com.task.paginationlist.data.utils.ILoader
import com.task.paginationlist.presentation.piclist.interfaces.OnPickClickListener
import com.task.paginationlist.presentation.piclist.views.activities.interfaces.IPaginatorAdapter
import com.task.paginationlist.presentation.piclist.views.adapters.holders.LoadingHolder
import com.task.paginationlist.presentation.piclist.views.adapters.holders.PicViewHolder
import com.task.paginationlist.presentation.piclist.views.paginator.Config

import java.text.ParseException

class PicAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>(), IPaginatorAdapter {

    private var isSuccessful = true
    private var reload: View.OnClickListener? = null
    private var onPicClick: OnPickClickListener? = null
    private var isFooterVisible: Boolean = false

    var loader: ILoader = PaginationListApplication.component!!.loader

    private var wallpapers: MutableList<WallpaperDb>? = null

    fun addData(models: MutableList<WallpaperDb>?, isRefreshing: Boolean) {
        if (isRefreshing) {
            this.wallpapers = null
        }
        if (models != null) {
            if (this.wallpapers == null) {
                this.wallpapers = models
            } else {
                this.wallpapers!!.addAll(models)
            }
        }
        isSuccessful = true
        notifyDataSetChanged()
    }

    override fun setReload(listener: View.OnClickListener) {
        this.reload = listener
    }

    override fun setOnPicClick(onPicClick: OnPickClickListener) {
        this.onPicClick = onPicClick
    }

    override fun setSuccessful(successful: Boolean) {
        isSuccessful = successful
        notifyItemChanged(itemCount - 1)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (isFooter(position)) {
            val loadingHolder = holder as LoadingHolder
            reload?.let {
                loadingHolder.loaderView!!.setReloadClick(it)
            }
            loadingHolder.loaderView!!.visibility = if (isFooterVisible) View.VISIBLE else View.GONE
            if (!isSuccessful) {
                loadingHolder.loaderView!!.showError()
            } else {
                holder.loaderView!!.showBar()
            }
        } else {
            val picHolder = holder as PicViewHolder
            loader.loadImage(wallpapers!![position].thumbUrl!!, picHolder.image)
            picHolder.image.setOnClickListener { v ->
                if (onPicClick != null) {
                    onPicClick!!.onPickClick(v, position, wallpapers!![position].id, wallpapers!!.size)
                }
            }
            var text = ""
            try {
                wallpapers!![position].creationDate?.let {
                    text = Config.formatDate(it)
                }

            } catch (e: ParseException) {
                Log.e(TAG, e.message)
            }

            picHolder.date.text = text
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == FOOTER_TYPE)
            LoadingHolder(LayoutInflater.from(parent.context).inflate(R.layout.footer_item, parent, false))
        else
            PicViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.pic_list_item, parent, false))
    }

    override fun getItemViewType(position: Int): Int {
        return if (isFooter(position)) FOOTER_TYPE else ITEM_TYPE
    }

    override fun setFooterVisibility(isAllLoaded: Boolean) {
        isFooterVisible = !isAllLoaded
    }

    private fun isFooter(position: Int): Boolean {
        return position == itemCount - 1
    }

    override fun realItemCount(): Int {
        return itemCount - 1
    }

    override fun getItemCount(): Int {
        return if (wallpapers != null) wallpapers!!.size + 1 else 1
    }

    companion object {

        private val TAG = "PicAdapter"
        val ITEM_TYPE = 1
        val FOOTER_TYPE = 0
    }
}
